### PR TITLE
Fix #5014: Implement java.io.BufferedWriter

### LINKED ISSUE #5014: Support java.io.BufferedWriter
While trying to cross-compile some zio-json code for supporting streaming json-lines, I ran into the issue that `java.io.BufferedWriter` was not implemented in Scala-js. 
I was also cross-compiling to Scala-native and there the BufferedWriter class is implemented. I tried to use the same implementation for Scala-js and it turns out it just works out of the box. This is the implementation: https://github.com/scala-native/scala-native/blob/d5d65e755a4fb40a1c562d50b046384000232b23/javalib/src/main/scala/java/io/BufferedWriter.scala#L4

Can this be included in Scala-js?
I could make a PR with this single file and include e.g. this test: https://github.com/scala-native/scala-native/blob/d5d65e755a4fb40a1c562d50b046384000232b23/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/io/BufferedWriterTest.scala#L10
Anything else?

### PR DESCRIPTION
**Pull request by ThijsBroersen**
Implementation copied from Scala-native

I excluded any copyright headers Scala-js uses in all files because this code comes from Scala-native

### CODE REVIEW

**Comment by ThijsBroersen** [comment]
Completed the CLA form

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    out.write(buffer, 0, pos)
+    out.flush()
+    pos = 0
+  }
+
+  def newLine(): Unit =
+    write(System.lineSeparator(), 0, System.lineSeparator().length)
+
+  override def write(c: Int): Unit =
+    write(Array(c.toChar), 0, 1)
```
Comment (gzm0) [inline javalib/src/main/scala/java/io/BufferedWriter.scala:30]: IIUC there is always at least one spot free in the buffer (otherwise it would be flushed).

So couldn't this be:

```scala
buffer(pos) = c.toChar
pos += 1
if (pos == len)
  flush()
```

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  override def write(s: String, off: Int, len: Int): Unit =
+    write(s.toCharArray, off, len)
+
+  def write(cbuf: Array[Char], off: Int, len: Int): Unit = {
+    ensureOpen()
+    val available = sz - pos
+    if (available >= len) {
+      System.arraycopy(cbuf, off, buffer, pos, len)
+      pos += len
+      if (pos == sz) flush()
```
[the remark this answers is not in the document]
Comment (gzm0) [reply javalib/src/main/scala/java/io/BufferedWriter.scala:41]: Also (IIRC) Scala.js style guide dictates that side-effecting calls are on their own line.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    val writer = new OutputStreamWriter(new ByteArrayOutputStream)
+    assertThrows(
+      classOf[IllegalArgumentException],
+      new BufferedWriter(writer, -1)
+    )
+  }
+
+  @Test def canWriteSmallChunksToBufferedWriter(): Unit = {
+    val stream = new ByteArrayOutputStream
+    val writer = new BufferedWriter(new OutputStreamWriter(stream))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala:22]: Have you considered a `StringWriter`?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package org.scalajs.testsuite.javalib.io
+
+import java.io._
+
+import org.junit.Test
+import org.junit.Assert._
+
+import org.scalajs.testsuite.utils.AssertThrows.assertThrows
+
+class BufferedWriterTest {
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala:10]: We'll probably have to add more tests, notably (probably not a complete list, sorry):
- operations on closed writers throw
- index out of bounds use cases
- corner cases pointed out in the implementation review (notably flushing behavior)
- writing to a buffered writer multiple times (with different buffer boundary conditions).

**Review by gzm0 — CHANGES_REQUESTED** [review]
TY for starting this!

I've done a first pass. As usual, even these relatively simple APIs turn out to be quite subtle :-/

**Comment by ThijsBroersen** [comment]
@gzm0 thanks for your extensive review of this code. As I mentioned the code is copied from Scala-native, it is 1:1 the same. I will go through your comments and see if the code can be improved. But I have some concern that if we make changes, the behaviour could be different than that of Scala-native. I would be nice if the can both benefit from the same implementation. Would it be possible Scala-js and Scala-native somehow share sources for these things?

**Comment by sjrd** [comment]
> But I have some concern that if we make changes, the behaviour could be different than that of Scala-native.

That doesn't matter. The reference is the JVM implementation. Both Scala.js and Scala Native must follow the specced behavior of the JDK. Scala.js, being more mature, is more demanding in terms of adhering to it.

> It would be nice if the can both benefit from the same implementation. Would it be possible Scala-js and Scala-native somehow share sources for these things?

No, that's not possible nor desirable. Copy-pasting is the way to go. Usually copying from JS to Native will be fine. Copying in the other direction tends to meet the higher quality standards of Scala.js, as you are experiencing now.

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
+package java.io
```
[the remark this answers is not in the document]
Comment (ThijsBroersen) [reply javalib/src/main/scala/java/io/BufferedWriter.scala:15]: done

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  def close(): Unit = if (!closed) {
+    flush()
+    out.close()
+    closed = true
+  }
+
+  def flush(): Unit = {
+    ensureOpen()
+    out.write(buffer, 0, pos)
+    out.flush()
```
[the remark this answers is not in the document]
Comment (ThijsBroersen) [reply javalib/src/main/scala/java/io/BufferedWriter.scala]: I read

> Thus one flush() invocation will flush all the buffers in a chain of Writers and OutputStreams

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  def newLine(): Unit =
+    write(System.lineSeparator(), 0, System.lineSeparator().length)
+
+  override def write(c: Int): Unit =
+    write(Array(c.toChar), 0, 1)
+
+  override def write(s: String, off: Int, len: Int): Unit =
+    write(s.toCharArray, off, len)
+
+  def write(cbuf: Array[Char], off: Int, len: Int): Unit = {
```
[the remark this answers is not in the document]
Comment (ThijsBroersen) [reply javalib/src/main/scala/java/io/BufferedWriter.scala:55]: I reimplemented the method

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    ensureOpen()
+    val available = sz - pos
+    if (available >= len) {
+      System.arraycopy(cbuf, off, buffer, pos, len)
+      pos += len
+      if (pos == sz) flush()
+    } else {
+      write(cbuf, off, available)
```
[the remark this answers is not in the document]
Comment (ThijsBroersen) [reply javalib/src/main/scala/java/io/BufferedWriter.scala]: I don't think it needs recursion. Reimplemented it to where it write to the buffer or it flushes the buffer and writes the all to the chained writer (as this is the behaviour I interpret from the JDK spec).

> Ordinarily this method stores characters from the given array into this stream's buffer, flushing the buffer to the underlying stream as needed. If the requested length is at least as large as the buffer, however, then this method will flush the buffer and write the characters directly to the underlying stream. Thus redundant BufferedWriters will not copy data unnecessarily.

File: javalib/src/main/scala/java/io/BufferedWriter.scala [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    pos = 0
+  }
+
+  def newLine(): Unit =
+    write(System.lineSeparator(), 0, System.lineSeparator().length)
+
+  override def write(c: Int): Unit =
+    write(Array(c.toChar), 0, 1)
+
+  override def write(s: String, off: Int, len: Int): Unit =
```
[the remark this answers is not in the document]
Comment (ThijsBroersen) [reply javalib/src/main/scala/java/io/BufferedWriter.scala:52]: It now writes nothing when these negative inputs occur

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala (hunk 3 above)
Comment (ThijsBroersen) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala]: now I have, keeps it simpler 🙃

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala (hunk 4 above)
Comment (ThijsBroersen) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/BufferedWriterTest.scala:24]: I have added more test cases.

**Comment by ThijsBroersen** [comment]
I have done another iteration, ready to review again.

**Comment by ThijsBroersen** [comment]
did sign the CLA (again)

**Comment by ThijsBroersen** [comment]
The AppVeyer CI seems be pass but if I check the logs than I see many 'failed'....
The Jenkins CI shows failed tests but all tests pass 'on-my-machine'

**Comment by sjrd** [comment]
The Jenkins CI jobs show the command that failed at the end of the log. In this case it seems the JVM test suite is failing:
```
Command was: sbt ++2.12.19 testSuiteJVM2_12/test testSuiteExJVM2_12/test
```
You should be able to replicate locally with `testSuiteJVM2_12/test`.

**Comment by ThijsBroersen** [comment]
Interesting, so the JDK spec says write nothing where the JVM implementation throws an StringIndexOutOfBoundsException exception

**Comment by sjrd** [comment]
If the JVM implementation is in direct contradiction with its (latest) spec, it might be a bug on the JVM. Consider searching the [JDK bug database](https://bugs.openjdk.org/projects/JDK/issues/JDK-8329348?filter=allopenissues). However, recall that you are not allowed to even *look at* the JDK implementation nor their tests.. If it's a known issue on the JVM we can exclude the test on the JVM with a comment to the JDK bug.

**Comment by ThijsBroersen** [comment]
Looks like it the was was reported before and 'fixed', so JVM should not throw for negative values but it still does 🤔 
https://bugs.openjdk.org/browse/JDK-8029804?jql=text%20~%20%22BufferedWriter%22

**Comment by ThijsBroersen** [comment]
But if we suspect a bug in the JVM code we can also not look at the source code and try to identify/fix it? Sounds silly...

**Comment by ThijsBroersen** [comment]
The JVM test also deviates in writing contents largen than the (available) buffer partially to the underlying stream. So it unnecessary splits/copies data. The is not what I interpret from the docs. 
e.g. if a test tries to write "Hello" to a BufferedWriter with size 3 on the JVM it writes "Hel" to the underlying stream and keeps "lo" in the buffer...

**Comment by ThijsBroersen** [comment]
I get the feeling the JDK is quite open to interpretation (or implemented wrong or not updated often) 🤣 . This also does make me a little sad 😢

**Comment by sjrd** [comment]
> I get the feeling the JDK is quite open to interpretation (or implemented wrong or not updated often) 🤣 . This also does make me a little sad 😢

This is actually quite rare. The JDK is remarkably well specified overall. If the latest JVM implementation still contradicts the spec, it might be worth opening a new issue, if only so that we can reference it in our tests to justify why we disagree with the JVM.

> But if we suspect a bug in the JVM code we can also not look at the source code and try to identify/fix it? Sounds silly...

Such is our life. You consider it silly; we see potential lawsuits if we don't play by the rules. We can report bugs, but not fix them nor investigate from the source. Essentially we have to treat the JDK as (free-as-in-beer) proprietary software.

**Comment by ThijsBroersen** [comment]
Silly from their perspective I mean, as in it discourages people from contributing. I understand the position of Scala-js.

**Comment by ThijsBroersen** [comment]
But for this PR.. will this differences be a blocker for now? Ignore those tests? Or should we replicate the JVM behaviour?

**Comment by sjrd** [comment]
It's not a blocker, but it should be documented as a JDK bug. See for example
https://github.com/scala-js/scala-js/blob/ede4dc1f70b416a3d3a5d4c27b64608656e77525/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala#L1695-L1715
and
https://github.com/scala-js/scala-js/blob/ede4dc1f70b416a3d3a5d4c27b64608656e77525/test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala#L1899-L1910
for examples.

**Comment by ThijsBroersen** [comment]
I think I will not continue this PR. I think it is better to skip to jdk classes and build pure Scala alternatives. It is also not satisfying that whatever was going to be implemented here would always be wrong because the spec is ambiguous or by not matching other openjdk implementations behaviour.
